### PR TITLE
Sample Course now has various autograding/TA grading assignments

### PR DESCRIPTION
**Pull request by holzbh**
Addresses most of #830. A new issue (#1164) has been opened for creating assignments with various late day usages.

### CODE REVIEW

File: .setup/bin/setup_sample_courses.py [hunk 1]
```diff
@@ -443,6 +446,185 @@ def create_gradeable_submission(src, dst):
         os.remove(zip_dst)
 
 
+def make_sample_json():
```
Comment (MasterOdin) [inline .setup/bin/setup_sample_courses.py]: So we're not always guaranteed to be building the sample course (if we only pass in say tutorial or whatever), I'd rather move the "make_sample_json()" function inside the course loop (right after the create function?) and make it a configuration option in the yml file (false by default).

Comment (holzbh) [inline .setup/bin/setup_sample_courses.py]: Can do, makes sense to me.

File: .setup/bin/setup_sample_courses.py [hunk 2]
```diff
@@ -443,6 +446,185 @@ def create_gradeable_submission(src, dst):
...
+    gradeables_percentages = []
+    gradeable_percentage_left = 100 - len(gradeables)
+    for i in range(len(gradeables)):
+        gradeables_percentages.append(random.randint(1, gradeable_percentage_left) + 1)
+        gradeable_percentage_left -= (gradeables_percentages[-1] - 1)
+    if gradeable_percentage_left > 0:
+        gradeables_percentages[-1] += gradeable_percentage_left
+
+    # Compute totals and write out each syllabus bucket in the "gradeables" field of customization.json
+    # TODO: Figure out what to do about "none" bucket items - does Rainbow Grades support this?
```
Comment (holzbh) [inline .setup/bin/setup_sample_courses.py]: It's a real bucket yes, but I don't think I've seen people type in

{
type: none,
count: 3,
percent 0.25,
[{id: .... }]
}

The syllabus bucket in the DB previously had nothing to do with how rainbow grades handled the type, but going forward as we offer a tool for generating customization files automatically, we would like to start using the syllabus buckets.

File: .setup/bin/setup_sample_courses.py [hunk 3]
```diff
@@ -443,6 +446,185 @@ def create_gradeable_submission(src, dst):
...
+    bucket_no = 0
+    for bucket,g_list in gradeables.items():
+        gradeables_json_output += "    {\n" +\
```
Comment (holzbh) [inline .setup/bin/setup_sample_courses.py]: Right, forgot about just dumping using json.dump() will adapt everywhere I'm currently writing out the whole mess by hand.

File: .setup/bin/setup_sample_courses.py [hunk 4]
```diff
@@ -489,7 +671,7 @@ def __init__(self, user):
         if 'rotating_section' in user:
             self.rotating_section = int(user['rotating_section'])
         if 'grading_registration_section' in user:
-            self.grading_registration_section = int(user['grading_registration_section'])
+            self.grading_registration_section = user['grading_registration_section']
```
Comment (holzbh) [inline .setup/bin/setup_sample_courses.py]: I added support for multiple grading registration sectiosn. By having the typecast we were limiting TAs to a max of 1 section per course, but this is too inflexible.

File: tests/rainbowGrades/test_sample.py [hunk 5]
```diff
@@ -122,7 +122,7 @@ def sample_rainbow_grades_test():
                     make_file.write("RAINBOW_GRADES_DIRECTORY=" + rainbow_tmp + "\n")
                 elif len(line) >= 18 and line[:18] == "REPORTS_DIRECTORY=":
                     make_file.write(os.path.join("REPORTS_DIRECTORY=__INSTALL__FILLIN__SUBMITTY_DATA_DIR__", "courses",
-                                                 "s17", "sample", "reports") + "\n")
+                                                 "f17", "sample", "reports") + "\n")
```
Comment (holzbh) [inline tests/rainbowGrades/test_sample.py]: Can do.

File: .setup/bin/setup_sample_courses.py (hunk 2 above)
Comment (MasterOdin) [inline .setup/bin/setup_sample_courses.py]: Okay.

File: .setup/bin/setup_sample_courses.py (hunk 4 above)
Comment (MasterOdin) [inline .setup/bin/setup_sample_courses.py]: Okay, yeah, you do the typecast later when you actually use this field. Got it.

File: .setup/bin/setup_sample_courses.py (hunk 2 above)
Comment (holzbh) [inline .setup/bin/setup_sample_courses.py]: Reading over this, I think you were misinterpreting my comment. I wasn't considering what to do if there is no bucket type, but more can Rainbow Grades handle "None (for practice only)".

As it turns out, Rainbow Grades [https://github.com/Submitty/Submitty/blob/master/RainbowGrades/main.cpp#L283-L298](will) look for anything starting with "none" at the moment (as well as a few other parsing oddities). So I can remove the comment.

**Comment by holzbh** [comment]
* Did the changes requested in the reviews

* Moved the JSON generation to be a method of Courses ( `make_course_json()` ) and made it work for any course.

* Extended courses.yml to support a flag for generation of customization.json per course

* Fixed some error reporting in test_sample.py, moved it to use the generated customization_sample.json

* Updated the expected Rainbow Grades output in the test to reflect TA names being used and course name being used.

**Pull request merged**